Workflow failure Quarkus IT - StartCommandDistTest#testWarningWhenOverridingBuildOptionsDuringStart

### Before reporting an issue

- [X] I have read and understood the above terms for submitting issues, and I understand that my issue may be closed without action if I do not follow them.

### Area

ci

### Describe the bug

Reference GHA run: https://github.com/keycloak/keycloak/actions/runs/7848451978/job/21419988110

New unreported flaky tests, probably due to inability to start JGroups Channel due to unavailable ports.

```
2024-02-09 19:41:58,884 ERROR [org.infinispan.CONFIG] (keycloak-cache-init) ISPN000660: DefaultCacheManager start failed, stopping any running components: org.infinispan.commons.CacheException: Unable to start JGroups Channel
-- 	at org.infinispan.remoting.transport.jgroups.JGroupsTransport.startJGroupsChannelIfNeeded(JGroupsTransport.java:604)
-- 	at org.infinispan.remoting.transport.jgroups.JGroupsTransport.start(JGroupsTransport.java:477)
-- 	at org.infinispan.remoting.transport.jgroups.CorePackageImpl$2.start(CorePackageImpl.java:63)
-- 	at org.infinispan.remoting.transport.jgroups.CorePackageImpl$2.start(CorePackageImpl.java:49)
-- 	at org.infinispan.factories.impl.BasicComponentRegistryImpl.invokeStart(BasicComponentRegistryImpl.java:616)
-- 	at org.infinispan.factories.impl.BasicComponentRegistryImpl.doStartWrapper(BasicComponentRegistryImpl.java:607)
-- 	at org.infinispan.factories.impl.BasicComponentRegistryImpl.startWrapper(BasicComponentRegistryImpl.java:576)
-- 	at org.infinispan.factories.impl.BasicComponentRegistryImpl$ComponentWrapper.running(BasicComponentRegistryImpl.java:807)
-- 	at org.infinispan.factories.impl.BasicComponentRegistryImpl.startDependencies(BasicComponentRegistryImpl.java:634)
-- 	at org.infinispan.factories.impl.BasicComponentRegistryImpl.doStartWrapper(BasicComponentRegistryImpl.java:598)
-- 	at org.infinispan.factories.impl.BasicComponentRegistryImpl.startWrapper(BasicComponentRegistryImpl.java:576)
-- 	at org.infinispan.factories.impl.BasicComponentRegistryImpl$ComponentWrapper.running(BasicComponentRegistryImpl.java:807)
-- 	at org.infinispan.factories.GlobalComponentRegistry.preStart(GlobalComponentRegistry.java:284)
-- 	at org.infinispan.factories.AbstractComponentRegistry.start(AbstractComponentRegistry.java:250)
-- 	at org.infinispan.manager.DefaultCacheManager.internalStart(DefaultCacheManager.java:779)
-- 	at org.infinispan.manager.DefaultCacheManager.start(DefaultCacheManager.java:747)
-- 	at org.infinispan.manager.DefaultCacheManager.<init>(DefaultCacheManager.java:411)
-- 	at org.keycloak.quarkus.runtime.storage.legacy.infinispan.CacheManagerFactory.startCacheManager(CacheManagerFactory.java:104)
-- 	at java.base/java.util.concurrent.FutureTask.run(FutureTask.java:264)
-- 	at java.base/java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1136)
-- 	at java.base/java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:635)
-- 	at java.base/java.lang.Thread.run(Thread.java:840)
-- Caused by: java.lang.IllegalStateException: fv-az1258-409-39315: failed to find an available port in ports [49664, 49665, 49666, 49667]
-- 	at org.jgroups.protocols.FD_SOCK2.createServer(FD_SOCK2.java:326)
-- 	at org.jgroups.protocols.FD_SOCK2.start(FD_SOCK2.java:161)
-- 	at org.jgroups.stack.ProtocolStack.startStack(ProtocolStack.java:905)
-- 	at org.jgroups.JChannel.startStack(JChannel.java:922)
-- 	at org.jgroups.JChannel._preConnect(JChannel.java:800)
-- 	at org.jgroups.JChannel.connect(JChannel.java:323)
-- 	at org.jgroups.JChannel.connect(JChannel.java:317)
-- 	at org.infinispan.remoting.transport.jgroups.JGroupsTransport.startJGroupsChannelIfNeeded(JGroupsTransport.java:602)
-- 	... 21 more
```

### Version

latest

### Expected behavior

.

### Actual behavior

.

### How to Reproduce?

.

### Anything else?

.

--- a/quarkus/tests/junit5/src/main/java/org/keycloak/it/junit5/extension/CLITestExtension.java
+++ b/quarkus/tests/junit5/src/main/java/org/keycloak/it/junit5/extension/CLITestExtension.java
@@ -97,11 +97,11 @@ public class CLITestExtension extends QuarkusMainTestExtension {
         infinispanContainer = configureExternalInfinispan(context);
 
         if (distConfig != null) {
-            onKeepServerAlive(context.getRequiredTestMethod().getAnnotation(KeepServerAlive.class));
-
             if (dist == null) {
                 dist = createDistribution(distConfig, getStoreConfig(context), getDatabaseConfig(context));
             }
+
+            onKeepServerAlive(context.getRequiredTestMethod().getAnnotation(KeepServerAlive.class), true);
 
             copyTestProvider(context.getRequiredTestClass().getAnnotation(TestProvider.class));
             copyTestProvider(context.getRequiredTestMethod().getAnnotation(TestProvider.class));
@@ -163,10 +163,10 @@ public class CLITestExtension extends QuarkusMainTestExtension {
         }
     }
 
-    private void onKeepServerAlive(KeepServerAlive annotation) {
+    private void onKeepServerAlive(KeepServerAlive annotation, boolean setting) {
         if(annotation != null && dist != null) {
             try {
-                dist.setManualStop(true);
+                dist.setManualStop(setting);
             } catch (Exception cause) {
                 throw new RuntimeException("Error when invoking " + annotation, cause);
             }
@@ -177,12 +177,11 @@ public class CLITestExtension extends QuarkusMainTestExtension {
     public void afterEach(ExtensionContext context) throws Exception {
         DistributionTest distConfig = getDistributionConfig(context);
 
-        if (distConfig != null) {
-            if (distConfig.keepAlive()) {
-                dist.stop();
-            }
+        if (dist != null) {
+            onKeepServerAlive(context.getRequiredTestMethod().getAnnotation(KeepServerAlive.class), false);
+            dist.stop();
 
-            if (DistributionTest.ReInstall.BEFORE_TEST.equals(distConfig.reInstall())) {
+            if (distConfig != null && DistributionTest.ReInstall.BEFORE_TEST.equals(distConfig.reInstall())) {
                 dist = null;
             }
         }

--- a/quarkus/tests/junit5/src/main/java/org/keycloak/it/junit5/extension/DistributionTest.java
+++ b/quarkus/tests/junit5/src/main/java/org/keycloak/it/junit5/extension/DistributionTest.java
@@ -29,6 +29,9 @@ import org.junit.jupiter.api.extension.ExtendWith;
 public @interface DistributionTest {
 
     boolean debug() default false;
+    /**
+     * If the distribution should be left running after the launch.
+     */
     boolean keepAlive() default false;
     boolean enableTls() default false;
 

--- a/quarkus/tests/junit5/src/main/java/org/keycloak/it/junit5/extension/KeepServerAlive.java
+++ b/quarkus/tests/junit5/src/main/java/org/keycloak/it/junit5/extension/KeepServerAlive.java
@@ -17,13 +17,10 @@
 
 package org.keycloak.it.junit5.extension;
 
-import org.keycloak.it.utils.KeycloakDistribution;
-
 import java.lang.annotation.ElementType;
 import java.lang.annotation.Retention;
 import java.lang.annotation.RetentionPolicy;
 import java.lang.annotation.Target;
-import java.util.function.Consumer;
 
 /**
  * {@link KeepServerAlive} is used in a distributiontest to keep the server alive on test / method level.

--- a/quarkus/tests/junit5/src/main/java/org/keycloak/it/utils/DockerKeycloakDistribution.java
+++ b/quarkus/tests/junit5/src/main/java/org/keycloak/it/utils/DockerKeycloakDistribution.java
@@ -52,7 +52,7 @@ public final class DockerKeycloakDistribution implements KeycloakDistribution {
         this.envVars.put(name, value);
     }
 
-    private GenericContainer getKeycloakContainer() {
+    private GenericContainer<?> getKeycloakContainer() {
         File distributionFile = new File("../../dist/" + File.separator + "target" + File.separator + "keycloak-" + Version.VERSION + ".tar.gz");
 
         if (!distributionFile.exists()) {
@@ -77,7 +77,7 @@ public final class DockerKeycloakDistribution implements KeycloakDistribution {
             image = new RemoteDockerImage(DockerImageName.parse("quay.io/keycloak/keycloak"));
         }
 
-        return new GenericContainer(image)
+        return new GenericContainer<>(image)
                 .withEnv(envVars)
                 .withExposedPorts(8080)
                 .withStartupAttempts(1)
@@ -113,6 +113,7 @@ public final class DockerKeycloakDistribution implements KeycloakDistribution {
             LOGGER.warn("Failed to start Keycloak container", cause);
         } finally {
             if (!manualStop) {
+                stop();
                 envVars.clear();
             }
         }
@@ -181,7 +182,6 @@ public final class DockerKeycloakDistribution implements KeycloakDistribution {
     private void cleanupContainer() {
         if (containerId != null) {
             try {
-                final String finalContainerId = containerId;
                 Runnable reaper = new Runnable() {
                     @Override
                     public void run() {
@@ -253,7 +253,7 @@ public final class DockerKeycloakDistribution implements KeycloakDistribution {
         }
 
         if (type.isInstance(this)) {
-            return (D) this;
+            return type.cast(this);
         }
 
         throw new IllegalArgumentException("Not a " + type + " type");

